### PR TITLE
[feature] Hybrid SAC

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

Hybrid SAC with the combined Q function (continuous as input, discrete as output). 

TODO: 
- Loss functions - do we need ratio between continuous and discrete? Or separate updates?

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 1]
```diff
@@ -115,24 +110,32 @@ def __init__(self, policy: TorchPolicy, trainer_params: TrainerSettings):
         )
         self.soft_update(self.policy.actor_critic.critic, self.target_network, 1.0)
 
+        _total_act_size = 0
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:113]: The ActionSpec (and so maybe the policy) has an `action_size` property to be backwards compatible with TF but also can be used for this.

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 2]
```diff
@@ -167,6 +170,17 @@ def _move_to_device(self, device: torch.device) -> None:
         self.target_network.to(device)
         self.value_network.to(device)
 
+    def _split_discrete_continuous(
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:173]: Lets make this a utility func

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 3]
```diff
@@ -257,8 +272,9 @@ def sac_value_loss(
                 )
                 value_losses.append(value_loss)
         else:
+            disc_log_probs = log_probs[:, self.policy.continuous_act_size :]
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:275]: We should be able to accomplish this with a `zip(entropy_coeffs, log probs)` of some sort

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py (hunk 2 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/sac/optimizer_torch.py:200]: I plan to remove this in favor of Tuples

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 4]
```diff
@@ -27,19 +26,13 @@ def __init__(
             stream_names: List[str],
             observation_shapes: List[Tuple[int, ...]],
             network_settings: NetworkSettings,
-            action_spec: ActionSpec,
+            continuous_act_size: int,
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:29]: Is this a hold out from the previous PR?

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 5]
```diff
@@ -27,19 +26,13 @@ def __init__(
...
+            discrete_act_size: int,
         ):
             super().__init__()
-            self.action_spec = action_spec
-            if self.action_spec.is_continuous():
-                self.act_size = self.action_spec.continuous_size
-                num_value_outs = 1
-                num_action_ins = self.act_size
+            num_value_outs = max(discrete_act_size, 1)
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:33]:
```suggestion
            num_value_outs = max(sum(self.policy.action_spec.discrete_branches), 1)
```

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 6]
```diff
@@ -27,19 +26,13 @@ def __init__(
...
+            num_action_ins = int(continuous_act_size)
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:34]: Use ActionSpec

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 7]
```diff
@@ -126,12 +118,14 @@ def __init__(self, policy: TorchPolicy, trainer_params: TrainerSettings):
...
         }
+        self._action_spec = self.policy.action_spec
 
         self.value_network = TorchSACOptimizer.PolicyValueNetwork(
             self.stream_names,
             self.policy.behavior_spec.observation_shapes,
             policy_network_settings,
-            self.policy.behavior_spec.action_spec,
+            self._action_spec.continuous_size,
+            sum(self._action_spec.discrete_branches),
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:128]: Just pass action spec?

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 8]
```diff
@@ -143,24 +137,32 @@ def __init__(self, policy: TorchPolicy, trainer_params: TrainerSettings):
             self.policy.actor_critic.critic, self.target_network, 1.0
         )
 
+        _total_act_size = 0
+        if self._action_spec.continuous_size > 0:
+            _total_act_size += self._action_spec.continuous_size
+        _total_act_size += sum(self._action_spec.discrete_branches)
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:143]: This can be accomplished with just `self._action_spec.continuous_size + sum(self._action_spec.discrete_branches)`

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py (hunk 2 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/sac/optimizer_torch.py:200]: Yea this comment was from before the addition of AgentActions. It can be totally removed.

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py (hunk 4 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/sac/optimizer_torch.py]: @ervteng I think we should pass the ActionSpec here.

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 9]
```diff
@@ -301,6 +316,13 @@ def sac_value_loss(
                     v_backup = min_policy_qs[name] - torch.mean(
                         branched_ent_bonus, axis=0
                     )
+                    # Add continuous entropy bonus to minimum Q
+                    if self._action_spec.continuous_size > 0:
+                        torch.sum(
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:321]: This code is repeated from line 295. Can it be consolidated?
